numpy/core/src/umath/dispatching.c

```diff
@@ -0,0 +1,697 @@
+ * (Based on `isinstance()`, the knowledge that non-abstract DTypes cannot
+ * be subclassed is used, however.)
+ *
+ * @param ufunc
+ * @param op_dtypes The DTypes that are either passed in (defined by an
+ *        operand) or defined by the `signature` as also passed in as
+ *        `fixed_DTypes`.
+ * @param out_info Returns the tuple describing the best implementation
+ *        (consisting of dtypes and ArrayMethod or promoter).
+ *        WARNING: Returns a borrowed reference!
+ * @returns -1 on error 0 on success.  Note that the output can be NULL on
+ *          success if nothing is found.
+ */
+static int
+resolve_implementation_info(PyUFuncObject *ufunc,
+        PyArray_DTypeMeta *op_dtypes[], PyObject **out_info)
+{
+    int nin = ufunc->nin, nargs = ufunc->nargs;
+    /* Use new style type resolution has to happen... */
+    Py_ssize_t size = PySequence_Length(ufunc->_loops);
+    PyObject *best_dtypes = NULL;
+    PyObject *best_resolver_info = NULL;
+
+    for (Py_ssize_t res_idx = 0; res_idx < size; res_idx++) {
+        /* Test all resolvers  */
+        PyObject *resolver_info = PySequence_Fast_GET_ITEM(
+                ufunc->_loops, res_idx);
+        PyObject *curr_dtypes = PyTuple_GET_ITEM(resolver_info, 0);
+        /*
+         * Test if the current resolver matches, it could make sense to
+         * reorder these checks to avoid the IsSubclass check as much as
+         * possible.
+         */
+
+        npy_bool matches = NPY_TRUE;
+        /*
+         * NOTE: We check also the output DType.  In principle we do not
+         *       have to strictly match it (unless it is provided by the
+         *       `signature`).  This assumes that a (fallback) promoter will
+         *       unset the output DType if no exact match is found.
+         */
+        for (Py_ssize_t i = 0; i < nargs; i++) {
+            PyArray_DTypeMeta *given_dtype = op_dtypes[i];
+            PyArray_DTypeMeta *resolver_dtype = (
+                    (PyArray_DTypeMeta *)PyTuple_GET_ITEM(curr_dtypes, i));
+            assert((PyObject *)given_dtype != Py_None);
+            if (given_dtype == NULL && i >= nin) {
+                /* Unspecified out always matches (see below for inputs) */
+                continue;
+            }
+            if (given_dtype == resolver_dtype) {
+                continue;
+            }
+            if (!resolver_dtype->abstract) {
+                matches = NPY_FALSE;
+                break;
+            }
+            if (given_dtype == NULL) {
+                /*
+                 * If the (input) was not specified, this is a reduce-like
+                 * operation.  Some ufuncs may have non-trivial promotion
+                 * (e.g. add/multiply ensure high precision).
+                 * Continuing here matches promoters: those can deal with that.
+                 * If we allow this path for ArrayMethod, the person
+                 * registering will have it works sense for the ufunc, a
```

Comment:
??  Did you mean something like "must have made sure it makes sense"? I don't quite understand the example.

Reply:
Changing to:
```
                 * If the input was not specified, this is a reduce-like
                 * operation.  Some ufuncs may have non-trivial promotion
                 * (e.g. add/multiply ensure high precision).
                 * Continuing here allows a promoter to handle reduce-like
                 * operations explicitly.  E.g. upcasting boolean to integer.
                 * TODO: The `!resolver_dtype->abstract` currently ensures that
                 *       this is a promoter.  If we allow ArrayMethods to use
                 *       abstract DTypes, we may have to reject it here or the
                 *       ArrayMethod has to implement the reduce promotion.
```
there are probably still some more loose ends with the possibility of ArrayMethod's being registered for abstract DTypes (but generally, I think it may be useful).

Reply:
Nvm, seeing the next comment, it now is:
```
                 * If an input was not specified, this is a reduce-like
                 * operation: reductions use `(operand_DType, NULL, out_DType)`
                 * as they only have a single operand.  This allows special
                 * reduce promotion rules useful for example for sum/product.
                 * E.g. `np.add.reduce([True, True])` promotes to integer.
                 *
                 * Continuing here allows a promoter to handle reduce-like
                 * promotions explicitly if necessary.
                 * TODO: The `!resolver_dtype->abstract` currently ensures that
                 *       this is a promoter.  If we allow ArrayMethods to use
                 *       abstract DTypes, we may have to reject it here or the
                 *       ArrayMethod has to implement the reduce promotion.
```
puh :)